### PR TITLE
[SPARK-24951][SQL] Table valued functions should throw AnalysisException

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
Previously TVF resolution could throw IllegalArgumentException if the data type is null type. This patch replaces that exception with AnalysisException, enriched with positional information, to improve error message reporting and to be more consistent with rest of Spark SQL.

## How was this patch tested?
Updated the test case in table-valued-functions.sql.out, which is how I identified this problem in the first place.

### CODE REVIEW

**Comment by holdensmagicalunicorn** [comment]
@rxin, thanks! I am a bot who has found some folks who might be able to help with the review:@hvanhovell, @cloud-fan and @gatorsmile

**Comment by rxin** [comment]
cc @gatorsmile @ericl who originally wrote this.

**Comment by ericl** [comment]
LGTM

**Comment by rxin** [comment]
Jenkins, retest this please.

File: sql/core/src/test/resources/sql-tests/results/table-valued-functions.sql.out [hunk 1]
```diff
@@ -83,8 +83,13 @@ select * from range(1, null)
 -- !query 6 schema
 struct<>
 -- !query 6 output
-java.lang.IllegalArgumentException
-Invalid arguments for resolved function: 1, null
+org.apache.spark.sql.AnalysisException
+error: table-valued function range with alternatives:
```
Comment (dilipbiswal) [inline sql/core/src/test/resources/sql-tests/results/table-valued-functions.sql.out]: @rxin will it be better to display the function name within single quotes ?

Comment (rxin) [reply sql/core/src/test/resources/sql-tests/results/table-valued-functions.sql.out]: probably could go either way

Comment (dilipbiswal) [reply sql/core/src/test/resources/sql-tests/results/table-valued-functions.sql.out]: @rxin Ok.

**Review by gatorsmile — COMMENTED** [review]
LGTM

Thanks! Merged to master.

**Pull request closed without merging**